File: 2015/01/cpp/solution.cpp

#include <fstream>
#include <ios>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int part_1(vector<char> input) {
  int floor = 0;

  for (char ch : input) {
    if (ch == '(') {
      floor += 1;
    } else if (ch == ')') {
      floor -= 1;
    }
  }

  return floor;
}

int part_2(vector<char> input) {
  int floor = 0;

  for (int i = 0; i < input.size(); i++) {
    char ch = input[i];
    if (ch == '(') {
      floor += 1;
    } else if (ch == ')') {
      floor -= 1;
    }

    if (floor == -1) {
      return i + 1;
    }
  }
  return -1;
}

int main() {
  vector<char> input;
  fstream file("../input.txt", fstream::in);
  char ch;
  while (file >> noskipws >> ch) {
    input.push_back(ch);
  }
  file.close();

  cout << "Part 1: " << part_1(input) << endl;
  cout << "Part 2: " << part_2(input) << endl;

  return 0;
}

File: 2015/01/python/solution.py
with open("../input.txt", "r") as file:
    actions = file.read().strip()


def part_1(actions: str) -> int:
    floor = 0

    for c in actions:
        if c == "(":
            floor += 1
        elif c == ")":
            floor -= 1

    return floor


def part_2(actions: str) -> int:
    floor = 0
    for i in range(len(actions)):
        c = actions[i]
        if c == "(":
            floor += 1
        elif c == ")":
            floor -= 1
        if floor == -1:
            return i + 1
    return -1


if __name__ == "__main__":
    print(f"Part 1: {part_1(actions)}")
    print(f"Part 2: {part_2(actions)}")
